front/src/applications/operationalStudies/helpers/formatTimetableItemSummaries.ts

```diff
@@ -0,0 +1,125 @@
+): Map<TimetableItemId, TimetableItemWithDetails> => {
+  const relevantTimetableItems = compact(
+    timetableItemIds.map((timetableItemId) => rawTimetableItems.get(timetableItemId))
+  );
+
+  const [trainScheduleWithDetails, pacedTrainWithDetails] = relevantTimetableItems.reduce<
+    [TrainScheduleWithDetails[], PacedTrainWithDetails[]]
+  >(
+    (acc, timetableItem) => {
+      const rollingStock = rollingStocks.find((rs) => rs.name === timetableItem.rolling_stock_name);
+
+      const timetableItemSummary = rawSummaries.get(timetableItem.id);
+
+      if (!timetableItemSummary) return acc;
+
+      let notHonoredReason: TimetableItemWithDetails['notHonoredReason'];
+      if (timetableItemSummary.status === 'success') {
+        if (isTooFast(timetableItem, timetableItemSummary)) notHonoredReason = 'trainTooFast';
+        if (isScheduledPointsNotHonored(timetableItem, timetableItemSummary))
+          notHonoredReason = 'scheduleNotHonored';
+      }
+
+      const startTime = new Date(timetableItem.start_time);
+
+      const otherProps =
+        timetableItemSummary.status === 'success'
+          ? {
+              isValid: true,
+              arrivalTime: new Date(startTime.getTime() + timetableItemSummary.time),
+              duration: new Duration({ milliseconds: timetableItemSummary.time }),
+              pathLength: formatKmValue(timetableItemSummary.length, 'millimeters', 1),
+              mechanicalEnergyConsumed: jouleToKwh(timetableItemSummary.energy_consumption, true),
+              pathItemTimes: {
+                base: timetableItemSummary.path_item_times_base,
+                provisional: timetableItemSummary.path_item_times_provisional,
+                final: timetableItemSummary.path_item_times_final,
+              },
+            }
+          : {
+              isValid: false,
+              arrivalTime: null,
+              duration: null,
+              pathLength: '',
+              mechanicalEnergyConsumed: 0,
+              invalidReason:
+                timetableItemSummary.status === 'pathfinding_not_found' ||
+                timetableItemSummary.status === 'pathfinding_input_error'
+                  ? timetableItemSummary.error_type
+                  : timetableItemSummary.status,
+            };
+
+      const formattedItem = {
+        ...timetableItem,
+        name: timetableItem.train_name,
+        startTime,
+        stopsCount:
+          (timetableItem.schedule?.filter(
+            (step) => step.stop_for && Duration.parse(step.stop_for).ms > 0
+          ).length ?? 0) + 1, // +1 to take the final stop (destination) into account
+        speedLimitTag: timetableItem.speed_limit_tag ?? null,
+        labels: timetableItem.labels ?? [],
+        rollingStock,
+        scheduledPointsNotHonored: notHonoredReason !== undefined,
+        notHonoredReason,
+        ...otherProps,
+      };
+
+      if (isPacedTrainResultWithPacedTrainId(formattedItem)) {
```

Comment:
This passes a partially constructed train result to the function. In particular, `formattedItem` will not yet have a `paced` field, so the function will always return false here?

Seems safer to check the `timetableItem.id` instead of operating on an incomplete value.

Reply:
Yeah, it should work, but is still passing a partially constructed object to the function. What I mean is that it's invalid to have an object with a paced train ID but without a `paced` field. Using the object before it's fully constructed can lead to surprises, especially for functions accepting both train schedules and paced trains.

Not a big deal though, I'm fine with the status quo.

Reply:
Unfortunately this ended up being a pretty deep rabbit hole, and fixing all of the type issues with my suggestion would result in a lot less readable code. Let's keep it that way for now, even if the types aren't perfectly correct.